### ISSUE TITLE
Views constructor

### ISSUE THREAD

**Issue by Reckziegel** [Reckziegel/FFP#10]
Even if  `entropy_pooling()` works fine, the function is still shy in the package as a whole. Maybe if there were constructors for the views, this function could have a central spot here and in the CMA (#6).

**Comment by Reckziegel** [comment]
Views on the mean:
``` r
library(ffp)

ret <- diff(log(EuStockMarkets))
n <- nrow(ret) 
# prior probabilities
p <- rep(1 / n, n)

### View: DAX returns will be 0.001 ###

# Constraint probs to sum 1
Aeq <- matrix(1, nrow = 1, ncol = n) 
beq <- 1 

# Attatch your View
Aeq <- rbind(Aeq, t(ret[ , "DAX"]))
beq <- rbind(beq, 0.001)

ep <- ffp:::entropy_pooling(p = p, Aeq = Aeq, beq = beq)

# Mu for DAX is 0.001!
empirical_stats(ret, ep)
#> # A tibble: 6 x 5
#>   .stat      DAX      SMI       CAC     FTSE
#>   <fct>    <dbl>    <dbl>     <dbl>    <dbl>
#> 1 Mu     0.00100  0.00104  0.000711 0.000604
#> 2 Std    0.0102   0.00917  0.0110   0.00793 
#> 3 Skew  -0.319   -0.417   -0.132    0.0558  
#> 4 Kurt   8.30     7.87     5.18     5.55    
#> 5 VaR    0.0276   0.0255   0.0281   0.0206  
#> 6 CVaR   0.0358   0.0333   0.0352   0.0249
```

<sup>Created on 2021-06-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by Reckziegel** [comment]
Views on volatility:

``` r
library(ffp)

ret <- diff(log(EuStockMarkets))
n <- nrow(ret) 
# prior probabilities
p <- rep(1 / n, n)

### View: Volatility of FTSE will be twice the average

# Constraint probs to sum 1
Aeq <- matrix(1, nrow = 1, ncol = n) 
beq <- 1 

# Attatch your View
Aeq <- rbind(Aeq, t(ret ^ 2))
beq <- rbind(beq, as.matrix(2 * (colMeans(ret) ^ 2)))

ep <- ffp:::entropy_pooling(p = p, Aeq = Aeq, beq = beq)

plot(ep, type = 'l')
```

![](https://i.imgur.com/ptCG4sA.png)

<sup>Created on 2021-06-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by Reckziegel** [comment]
Views on the correlation structure: 

``` r
# data
ret <- diff(log(EuStockMarkets))
n <- nrow(ret) 
p <- rep(1 / n, n) # prior probabilities

# Views on Correlation
mu <- colMeans(ret)
sd <- apply(ret, 2, sd)

# panic that throws correlations to the roof!
co <- matrix(0.95, 4, 4) + runif(16, 0, 0.03)
diag(co) <- 1

# Constraint probs to sum 1
Aeq <- matrix(1, nrow = 1, ncol = n) 
beq <- 1 

# Attach the view on cor
for (k in 1:ncol(co)) {
  for (l in k:ncol(co)) {
    Aeq <- rbind(Aeq ,t(ret[ , k] * ret[ , l]))
    beq <- rbind(beq, mu[[k]] * mu[[l]] * sd[[k]] * sd[[l]] * co[k, l])
  }
}

# Numerical Entropy Pooling
objective <- function(v, p, Aeq, beq){
  x <- exp(log(p) - 1 - t(Aeq) %*% v)
  x <- apply(cbind(x, 10 ^ -32), 1, max)
  L <- t(x) %*% (log(x) - log(p) + t(Aeq) %*% v) - t(beq) %*% v
  -L
}
gradient <- function(v, p, Aeq, beq){
  x <- exp(log(p) - 1 - t(Aeq) %*% v)
  beq - Aeq %*% x
}

entropy_nlminb <- function(p, Aeq, beq) {
  v_dual <- stats::nlminb(
    start     = rep(0, NROW(Aeq)),
    objective = objective,
    gradient  = gradient,
    Aeq       = Aeq,
    beq       = beq,
    p         = p)
  v <- v_dual$par
  v <- v_dual$par
  p <- exp(log(p) - 1 - t(Aeq) %*% v)
  as.double(p / sum(p))
}

opt_nlminb <- entropy_nlminb(p = p, Aeq = Aeq, beq = beq)
plot(opt_nlminb, type = 'l')
```

![](https://i.imgur.com/dGR0jkc.png)

<sup>Created on 2021-06-30 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

NOTE: This is a more complicated problem and requires a different solver than `NlcOptim::solnl`.

**Comment by Reckziegel** [comment]
This issue will be totally addressed in the 0.2.0 release.
